atc/db/resource.go

```diff
@@ -966,9 +992,36 @@ func populateBuildSummary(buildSummary *atc.BuildSummary,
 	}
 
 	if lastCheckBuildPlan.Valid {
-		err := json.Unmarshal([]byte(lastCheckBuildPlan.String), &buildSummary.PublicPlan)
-		if err != nil {
-			return err
+		if atc.EnableGlobalResources {
+			// When global_resources is enabled, name in check plan could be any
+			// resource name under the same scope. So we need to extra the plan
+			// and replace name with the current resource's name.
+			var plan struct {
+				ID    string `json:"id,omitempty"`
+				Check struct {
+					Name           string          `json:"name"`
+					Type           string          `json:"type"`
+					Resource       string          `json:"resource,omitempty"`
+					ImageGetPlan   json.RawMessage `json:"image_get_plan,omitempty"`
+					ImageCheckPlan json.RawMessage `json:"image_check_plan,omitempty"`
+				} `json:"check,omitempty"`
+			}
```

Comment:
Ideally we wouldn't have to keep this in sync with the public plan - can we at least extract a shared type for it?

Reply:
I noticed that, even with this change, the GUI resource checking page might still show bad "check: resource name". So I reverted this change. Maybe we can address the issue after merging this PR.